src/Contract/Collection.php

```diff
@@ -192,7 +193,7 @@
  * @template-extends Lastable<TKey, T>
  * @template-extends Limitable<TKey, T>
  * @template-extends Linesable<TKey, T>
- * @template-extends Mapable<TKey, T>
+ * @template-extends Mapable<TKey, T, V>
```

Comment:
seeing the errors on this I'm thinking it's not the right way.

Looking at the [Doctrine ArrayCollection](https://github.com/doctrine/collections/blob/1.6.x/lib/Doctrine/Common/Collections/ArrayCollection.php#L347) the template variable is defined on the method.

I played with the Psalm example a bit more and came up with another version which works while keeping the template param at the same level:
https://psalm.dev/r/e840743851
https://phpstan.org/r/5166ee20-1783-4d1d-93f6-91e4ac3570b4

Note the callback is passed from the invoke method. I think the issue with the initial one might be that the template is defined at a different level from where the callback is passed, so it "can't know" the return type of the callback. This is just my assumption, I can't claim to understand this fully 😄

Reply:
Going to analyze this later in the day for sure!

Reply:
Voila, I updated the top message with the relevant PSalm snippets.

I'm still searching for the best way to implement this.